python_modules/dagster/dagster_tests/general_tests/test_record.py

```diff
@@ -600,3 +601,104 @@ class SubSub(Sub):
+
+
+def test_generic_with_propagate() -> None:
+    T = TypeVar("T")
+
+    class Base(Generic[T]): ...
+
+    @record
+    class RecordBase(Base[T]):
+        label: Optional[str] = None
+
+    @record
+    class SubAdditionalArg(RecordBase):
+        some_val: str
+
+    obj = SubAdditionalArg(some_val="hi")
+    assert SubAdditionalArg(some_val="hi").some_val == "hi"
+    assert copy(obj, label="...").label == "..."
+    assert copy(obj, some_val="new").some_val == "new"
+
+    @record
+    class SubAdditionalArgRecursive(RecordBase):
+        vals: Sequence[RecordBase]
+
+    obj = SubAdditionalArgRecursive(vals=[SubAdditionalArg(some_val="hi")])
+    assert len(obj.vals) == 1
+    assert copy(obj, label="...").label == "..."
+
+    @record
+    class SubAdditionalArgSpecific(RecordBase[int]):
+        vals: Sequence[RecordBase[str]]
+
+    obj = SubAdditionalArgSpecific(vals=[SubAdditionalArg(some_val="hi")])
+    assert len(obj.vals) == 1
+    assert copy(obj, label="...").label == "..."
+
+    @record
+    class SubAdditionalArgVariableBase(RecordBase[T]):
+        vals: Sequence[RecordBase[T]]
+        val: Base[T]
+
+    obj = SubAdditionalArgVariableBase(
+        vals=[SubAdditionalArg(some_val="hi")], val=SubAdditionalArg(some_val="bye")
+    )
+    assert len(obj.vals) == 1
+    assert copy(obj, label="...").label == "..."
+
+    @record
+    class SubSubVariableBaseSpecific(SubAdditionalArgVariableBase[str]): ...
+
+    obj = SubSubVariableBaseSpecific(
+        vals=[SubAdditionalArg(some_val="hi")], val=SubAdditionalArg(some_val="bye")
+    )
+    assert len(obj.vals) == 1
+    assert copy(obj, label="...").label == "..."
+
+    @record
+    class SubSubVariableBaseAny(SubAdditionalArgVariableBase): ...
+
+    obj = SubSubVariableBaseAny(
+        vals=[SubAdditionalArg(some_val="hi")], val=SubAdditionalArg(some_val="bye")
+    )
+    assert len(obj.vals) == 1
+    assert copy(obj, label="...").label == "..."
+
+
+def test_generic_with_propagate_type_checking() -> None:
+    T = TypeVar("T")
+
+    class Base(Generic[T]): ...
+
+    @record
+    class RecordBase(Base[T]):
+        inner: T
+
+    @record
+    class SpecificRecord(RecordBase):
+        val1: RecordBase
+        val2: RecordBase[str]
+        val3: Base
+        val4: Base[int]
+
+    valid_record = SpecificRecord(
+        inner=...,
+        val1=RecordBase(inner=1.23),
+        val2=RecordBase(inner="hi"),
+        val3=RecordBase(inner=1.23),
+        val4=RecordBase(inner=1),
+    )
+
+    with pytest.raises(CheckError, match='"val1" is not a RecordBase'):
+        copy(valid_record, val1=Base())
+
+    with pytest.raises(CheckError, match='"val2" is not a RecordBase'):
+        copy(valid_record, val2=Base())
+
+    with pytest.raises(CheckError, match='"val3" is not a Base'):
+        copy(valid_record, val3=3)
+
+    with pytest.raises(CheckError, match='"val4" is not a Base'):
+        copy(valid_record, val4=4)
```

Comment:
we should add a xfail test case with `@record_custom` subclass just to keep track